### PR TITLE
Rollback on failed unmount operations

### PR DESCRIPTION
**Pull request by kursatkobya**
CreateDir: Add a new class to track to hold directory creation backlog.
FileToolkitWithUndo: Use smart pointer instead of raw pointer.
Rollback on failed unmount operations
unittests for CreateDir class

Note : one tests marked to be skipped. It is necessary to mark it as "DISABLED" due to current CI conditions.

### CODE REVIEW

File: common/filetoolkitwithundo.cpp [hunk 1]
```diff
@@ -180,15 +123,33 @@ bool FileToolkitWithUndo::overlayMount(const std::string &lower,
...
+        || !m_create.createDirectory(upper)
+        || !m_create.createDirectory(work)
+        || !m_create.createDirectory(dst))
     {
+        m_create.rollBack();
         log_error() << "Failed to create lower/upper/work directory for overlayMount. lowerdir=" <<
                        lower << ", upperdir=" << upper << ", workdir=" << work;
         return false;
     }
+    m_create.clear();
```
Comment (tobsan) [inline common/filetoolkitwithundo.cpp]: I don't really like this solution at all. Wouldn't it make more sense if one could have a local FileToolKitWithUndo instance that also has a cleanup handler API?

```
FileToolkKitWithUndo ft;
if (!ft.createDirectory(lower) || ...) {
   // Something went wrong
} else 
    m_cleanupHandlers.emplace_back(ft);
}
```

Comment (tobsan) [reply common/filetoolkitwithundo.cpp]: In the went wrong step, the local ft would go out of scope and would automatically cleanup as part of its dtor.

File: common/filetoolkitwithundo.h [hunk 2]
```diff
@@ -127,14 +109,9 @@ class FileToolkitWithUndo
...
-private :
-    /**
-     * @brief createParentDirectory Recursively tries to create the directory pointed to by path.
-     * @param path The directory path to be created.
-     * @return true on success, false on failure
-     */
-    bool createParentDirectory(const std::string &path);
+    std::vector<std::unique_ptr<CleanUpHandler>> m_cleanupHandlers;
+
+    CreateDir m_create;
```
Comment (tobsan) [inline common/filetoolkitwithundo.h]: Separating the directory creation code like this feels unnatural, and causes code duplication later on (either clean, or add the stuff done with m_create to a cleanuphandler anyway)

File: common/unit-test/createdir_unittest.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+{
+public:
+
+    void SetUp() override
+    {
+    }
+
+    CreateDir cd;
+
+};
```
Comment (tobsan) [inline common/unit-test/createdir_unittest.cpp]: It would make more sense to test parts of the FileToolkitWithUndo instead, in line with my other comments.

File: common/unit-test/createdir_unittest.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    chmod("/tmp/test/error/in/nested", 444);
+    ASSERT_FALSE(cd.createDirectory("/tmp/test/error/in/nested/directory/error"));
+    ASSERT_TRUE(cd.rollBack());
+}
+
+TEST_F(CreateDirTest, tempError)
+{
+    EXPECT_THROW(cd.tempDir("tmp/sc-unappropriate/template"), SoftwareContainerError);
+    ASSERT_TRUE(cd.rollBack());
+}
```
Comment (tobsan) [inline common/unit-test/createdir_unittest.cpp]: These tests are good though, this is the type of stuff we want to test with the cleanuphandlers.

File: libsoftwarecontainer/src/container.cpp [hunk 5]
```diff
@@ -110,11 +110,16 @@ bool Container::initialize()
 {
     if (m_state < ContainerState::PREPARED) {
         std::string gatewayDir = gatewaysDir();
-        if (!createDirectory(gatewayDir)) {
+        if (!m_create.createDirectory(gatewayDir)) {
+            m_create.rollBack();
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp]: This change makes little sense as any parent directories would have been cleaned anyway when the container dtor is called (and this is kinda fatal error).

File: libsoftwarecontainer/src/container.cpp [hunk 6]
```diff
@@ -110,11 +110,16 @@ bool Container::initialize()
...
             log_error() << "Could not create gateway directory "
                         << gatewayDir << ": " << strerror(errno);
             return false;
         }
+        m_create.clear();
+        if (!pathInList(gatewayDir)) {
+            m_cleanupHandlers.emplace_back(new DirectoryCleanUpHandler(gatewayDir));
+        }
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp]: Here is another example of how this change creates double checks instead of using the existing FileToolkitWithUndo.

File: libsoftwarecontainer/src/container.cpp [hunk 7]
```diff
@@ -620,14 +627,40 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
...
     }
 
-    return bindMountCore(pathInHost, pathInContainer, tempPath, readOnly);
+    if (!bindMountCore(pathInHost, pathInContainer, tempPath, readOnly)) {
+        // bindMountInContainer operation is not successful. Clean mounted tempPath.
+        MountCleanUpHandler rollback{tempPath};
+        rollback.clean();
+        // clean createdDirectories as well
+        m_create.rollBack();
+        return false;
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp]: This is going in the right direction though, but it could be done nicer.

File: libsoftwarecontainer/src/container.cpp [hunk 8]
```diff
@@ -620,14 +627,40 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
...
+            m_cleanupHandlers.emplace_back(std::move(tempFileCleaner));
+        }
+        // clear rollback history
+        m_create.clear();
+    } else {
+        // add cleanup for tempPath file
+        m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
+    }
+    m_cleanupHandlers.emplace_back(new MountCleanUpHandler(tempPath));
+    return true;
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp]: This feels largely like we're copying parts of the code from the FileToolKitWithUndo and making it more specific. Do we have to?

File: libsoftwarecontainer/src/softwarecontainer.cpp [hunk 9]
```diff
@@ -440,11 +440,16 @@ void SoftwareContainer::checkWorkspace()
...
+        if(!m_create.createDirectory(rootDir)) {
+            m_create.rollBack();
             std::string message = "Failed to create container root directory";
             log_error() << message;
             throw SoftwareContainerError(message);
         }
+        m_create.clear();
+        if (!pathInList(rootDir)) {
+            m_cleanupHandlers.emplace_back(new DirectoryCleanUpHandler(rootDir));
+        }
```
Comment (tobsan) [inline libsoftwarecontainer/src/softwarecontainer.cpp]: This part of the code has nothing to do with failed mounts.

File: libsoftwarecontainer/src/container.cpp (hunk 5 above)
Comment (tobsan) [reply libsoftwarecontainer/src/container.cpp]: Also this has nothing to do with failing mounts.

File: common/createdir.h [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+     * @param path The directory path to be created.
+     * @return true on success, false on failure
+     */
+    bool createParentDirectory(const std::string path);
+
+    /**
+     * @brief m_createDirectoryCleanupHandlers holds cleanup handler for each created directory.
+     * The main purpose of this is to rollback in any errors.
+     */
+    std::vector<std::unique_ptr<DirectoryCleanUpHandler>> m_rollbackCleaners;
```
Comment (tobsan) [inline common/createdir.h]: Is there a reason for using pointers here and not objects or references directly?

File: common/createdir.h [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+     * @return A string path pointing to the newly created temporary directory.
+     */
+    std::string tempDir(std::string templatePath);
+
+    /**
+     * @brief m_tempFileCleaners holds cleanup handler for each created temporary directory.
+     * The main purpose of this is to rollback in any errors while bind-mounting. On success
+     * m_tempFileCleaners will be moved to cleanup handlers of FileToolkitWithUndo
+     */
+    std::vector<std::unique_ptr<DirectoryCleanUpHandler>> m_tempFileCleaners;
```
Comment (tobsan) [inline common/createdir.h]: Should this member really be public?

**Review by tobsan — CHANGES_REQUESTED** [review]
I find the CreateDir class completely superfluous. 

Furthermore, this PR does not solve the issues that might arise from a failing bind mount inside the container. What if the mount-move fails? What if a re-mount to read-only fails? These cases does not seem to be addressed here.

File: common/filetoolkitwithundo.cpp (hunk 1 above)
Comment (kursatkobya) [reply common/filetoolkitwithundo.cpp]: You mean in overlayMount method of FileToolkitWithUndo class should create a inception FileToolkKitWithUndo class. I don't believe that would be nicer way. It does not make more sense to me, no.

File: libsoftwarecontainer/src/container.cpp (hunk 5 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/container.cpp]: ok it will be removed.

File: common/createdir.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    return success;
+}
+
+std::string CreateDir::tempDir(std::string templ)
+{
+    char *dir = const_cast<char*>(templ.c_str());
+    dir = mkdtemp(dir);
+    if (dir == nullptr) {
+        std::string message = "Failed to create buffered Directory: " + std::string(strerror(errno));
```
Comment (tobsan) [inline common/createdir.cpp]: This isn't necessarily a buffered directory, it's just a tempdir.

File: common/filetoolkitwithundo.h [hunk 13]
```diff
@@ -127,14 +109,8 @@ class FileToolkitWithUndo
...
-    std::vector<CleanUpHandler *> m_cleanupHandlers;
-private :
-    /**
-     * @brief createParentDirectory Recursively tries to create the directory pointed to by path.
-     * @param path The directory path to be created.
-     * @return true on success, false on failure
-     */
-    bool createParentDirectory(const std::string &path);
+    std::vector<std::unique_ptr<CleanUpHandler>> m_cleanupHandlers;
+    std::vector<std::unique_ptr<CreateDir>> m_createDirList;
```
Comment (tobsan) [inline common/filetoolkitwithundo.h]: It would be good with some comments about why these are separated.

File: common/unit-test/createdir_unittest.cpp (hunk 4 above)
Comment (tobsan) [reply common/unit-test/createdir_unittest.cpp]: Comments on all these test cases would be good.

File: libsoftwarecontainer/src/container.cpp [hunk 14]
```diff
@@ -622,12 +625,23 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
                 log_error() << "Could not create file " << tempPath;
                 return false;
             }
-            m_cleanupHandlers.push_back(new FileCleanUpHandler(tempPath));
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp:625]: Why is this line removed? If I read this correctly, this was the cleanup handler for a file that was touched in case we didn't mount to a dir but to a file. So why not cleanup that file just as we clean up the directories? Especially in the case where things go bad.

File: libsoftwarecontainer/src/container.cpp [hunk 15]
```diff
@@ -622,12 +625,23 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
...
         }
     }
 
-    return bindMountCore(pathInHost, pathInContainer, tempPath, readOnly);
+    if (!bindMountCore(pathInHost, pathInContainer, tempPath, readOnly)) {
+        // bindMountInContainer operation is not successful. Clean mounted tempPath.
+        MountCleanUpHandler rollback{tempPath};
+        rollback.clean();
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp:634]: So we create a `MountCleanUpHandler` with `tempPath` and then instantly clear it of that path?

File: common/createdir.cpp [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    if (!pathInList(path)){
+        m_rollbackCleaners.emplace_back(path);
+    }
+    return true;
+}
+
+void CreateDir::clear()
+{
+    m_rollbackCleaners.clear();
+}
```
Comment (tobsan) [inline common/createdir.cpp]: This call is only used once in the code (and I'm not sure it should). Consider removing it.

**Review by tobsan — CHANGES_REQUESTED** [review]
Looks better, see further comments.

File: common/createdir.h [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+     * @param path Path of directory to be created
+     * @return true on success, false on failure
+     */
+    bool createDirectory(const std::string path);
+
+    /**
+     * @brief rollBack calls clean method for all members of m_createDirectoryCleanupHandlers.
+     *
+     * This function is created as a helper to callers of createDirectory functions. It is strongly
+     * recommended that this functions should be called if createDirectory function returns false.
```
Comment (breakreturn) [inline common/createdir.h]: I think comments like this also should be summarized, in the context of how to use this class, as a class comment above the class declaration.

File: common/createdir.h [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+class CreateDir {
+    LOG_DECLARE_CLASS_CONTEXT("CD", "Create Directory");
+public:
+    ~CreateDir();
+
+    /**
+     * @brief createDirectory Create a directory, and if successful append it
+     *  to a list of dirs to be deleted in the dtor. Since nestled dirs will
+     *  need to be deleted in reverse order to creation insert to the beginning
+     *  of the list.
```
Comment (breakreturn) [inline common/createdir.h]: Unnecessary implementation details here. Could we just say that "Creates a directory. If a directory is successfully created, it is removed when this object is destroyed"?

File: libsoftwarecontainer/src/container.cpp (hunk 15 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/container.cpp:634]: Yes, this is a line just for cleaning mount. Instead we can umount2 here, but i think it would be quite ugly.

File: common/createdir.h [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+     * This function is created as a helper to callers of createDirectory functions. It is strongly
+     * recommended that this functions should be called if createDirectory function returns false.
+     *
+     * @return true on success, false on failure
+     */
+    bool rollBack();
+
+    /*@brief clear removes all elements from the m_createDirectoryCleanupHandlers
+     *
+     * */
```
Comment (breakreturn) [inline common/createdir.h]: Formatting

File: common/createdir.h [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    void clear();
+
+    /**
+     * @brief tempDir Creates a uniquely named temporary directory according to templatePath.
+     *
+     * @warning The temporary path will be destroyed when the instance of FileToolkitWithUndo
+     *  is destroyed.
```
Comment (breakreturn) [inline common/createdir.h]: Which instance is this? When is it destroyed?

File: libsoftwarecontainer/src/container.cpp (hunk 15 above)
Comment (tobsan) [reply libsoftwarecontainer/src/container.cpp:634]: Ah yes, I read it as a CreateDir object. My bad.

File: common/createdir.h [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+private :
+    /**
+     * @brief createParentDirectory Recursively tries to create the directory pointed to by path.
+     * @param path The directory path to be created.
+     * @return true on success, false on failure
+     */
+    bool createParentDirectory(const std::string path);
+
+    /**
+     * @brief m_createDirectoryCleanupHandlers holds cleanup handler for each created directory.
```
Comment (breakreturn) [inline common/createdir.h]: Is this comment about an old naming for the below member? I think it's enough to comment on the member and let it be implicit that the comment is about the member directly below the comment, to avoid mistakes like this.

File: libsoftwarecontainer/src/container.cpp (hunk 14 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/container.cpp:625]: Here this is moved to end of the function. See the snippet

```
if (!pathIsDirectory) {
    m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
 }
```

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: common/createdir.h (hunk 18 above)
Comment (kursatkobya) [reply common/createdir.h]: Those detailes are inherited from previous implementation, will change as you requested.

File: common/createdir.h [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    bool rollBack();
+
+    /**
+     * @brief tempDir Creates a uniquely named directory according to templatePath.
+     *
+     * @warning The created directory will be destroyed when the object is destroyed or rollBack
+     * method is called.
+     *
+     * @param templ a template Path used to create the path of the temporary directory, including
```
Comment (breakreturn) [inline common/createdir.h]: Typo in the param

File: common/createdir.h [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+     *  XXXXXX which will be replaced with a unique ID for the temporary directory
+     *
+     * @throws SoftwareContainerError if template path is not appropriate.
+     * @return A string path pointing to the newly created temporary directory.
+     */
+    std::string tempDir(std::string templatePath);
```
Comment (breakreturn) [inline common/createdir.h]: When does one use createDirectory and tempDir? Does tempDir create a directory?

File: common/createdir.h [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+private :
+    /**
+     * @brief createParentDirectory Recursively tries to create the directory pointed to by path.
+     * @param path The directory path to be created.
+     * @return true on success, false on failure
+     */
+    bool createParentDirectory(const std::string path);
+
+    /**
+     * @brief m_rollbackCleaners holds cleanup handler for each created directory.
```
Comment (breakreturn) [inline common/createdir.h]: Why mention the variable name here again? It makes it easy to miss when updating the code.

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: common/unit-test/createdir_unittest.cpp (hunk 4 above)
Comment (tobsan) [reply common/unit-test/createdir_unittest.cpp]: None of these tests actually test that `rollBack()` removes the files, or that they do not already exist before running the tests.

Also, this is the only place where the `rollBack()` function is used as a public member. The tests could be rewritten so that `rollBack()` doesn't have to be public. Not really important though...

Comment (breakreturn) [reply common/unit-test/createdir_unittest.cpp]: @tobsan Isn't rollback() one of the main features? I'm probably misreading your comment...

File: libsoftwarecontainer/src/container.cpp [hunk 25]
```diff
@@ -622,12 +625,24 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
...
+        return false;
+    }
 
+    // bindMountInContainer succeed. Add cleanup for mounted tempPath
+    m_createDirList.push_back(std::move(createDirInstance));
+    if (!pathIsDirectory) {
+        m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
+    }
+
+    m_cleanupHandlers.emplace_back(new MountCleanUpHandler(tempPath));
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp:644]: You could probably rewrite this so that you use the same MountCleanupHandler as mentioned above (the one that cleans instantly), and then depending on result clean or save to `m_cleanupHandlers`.

File: common/unit-test/createdir_unittest.cpp (hunk 4 above)
Comment (tobsan) [reply common/unit-test/createdir_unittest.cpp]: @breakreturn yes, but rollback is done automatically upon destruction of the object.

File: common/createdir.h (hunk 23 above)
Comment (kursatkobya) [reply common/createdir.h]: tempDir and createDirectory are independent two functions. tempDir creates uniquely named temporary directory from template. 
Meanwhile createDirectory just creates a directory in given path. 
For more information one can look into bindMount function in FileToolkitWithUndo class.

File: common/unit-test/createdir_unittest.cpp (hunk 4 above)
Comment (breakreturn) [reply common/unit-test/createdir_unittest.cpp]: @tobsan Yeah, but I thought we needed that for some reason? Why would it otherwise exist at all on the public interface?

File: common/createdir.h (hunk 23 above)
Comment (breakreturn) [reply common/createdir.h]: So "tempDir" should probably be named in a similar fashion as "createDirectory" is then? Otherwise the user will be forced to wonder why one says "create" and the other not, while both logically should create something.

And what do you mean with look into FileToolkitWithUndo? How would the user know that?

File: libsoftwarecontainer/src/container.cpp (hunk 25 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/container.cpp:644]: I don't see any perk of doing like this. I prefer this way if it is not disruptive for you.

File: common/unit-test/createdir_unittest.cpp (hunk 4 above)
Comment (tobsan) [reply common/unit-test/createdir_unittest.cpp]: We need the functionality, but it can be triggered by just letting the object die, which will happen around the same time anyway in most cases. But it doesn't hurt to keep it.

Comment (kursatkobya) [reply common/unit-test/createdir_unittest.cpp]: We had one test examines `rollBack`. But as you know it is disabled for now due to root user issue on unit tests. Yet `rollBack` function can be removed as you mentioned.

Comment (breakreturn) [reply common/unit-test/createdir_unittest.cpp]: I think we should have as few methods as possible visible to the user and as crisp and clear interface possible as well. More methods means more ways to do the same thing, means more tests, means more ways to fail etc. I'd just like to see one clear explanation of how to use this class, if it's "might be needed but often happens at the same time", I think there is a risk it's not thought through.

File: common/createdir.h (hunk 23 above)
Comment (kursatkobya) [reply common/createdir.h]: I didn't create those naming but i believe it is not the worst naming. Still i will change it to createTemplateDir which make more sense. 

These functions are not for users of SoftwareContainer but for developers of it. So if FileToolkitWithUndo does not ring any bells then probably the one doesn't need to have any  further look in this class!

File: common/createdir.h [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    bool createDirectory(const std::string path);
+
+    /**
+     * @brief createTempDirectoryFromTemplate creates a uniquely named directory according to
+     * templatePath.
+     *
+     * @warning The created directory will be destroyed when the object is destroyed
+     *
+     * @param templatePath a template Path used to create the path of the temporary directory,
```
Comment (breakreturn) [inline common/createdir.h]: What's the template? What should one pass in here?

File: common/createdir.h [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+     * @throws SoftwareContainerError if template path is not appropriate.
+     * @return A string path pointing to the newly created temporary directory.
+     */
+    std::string createTempDirectoryFromTemplate(std::string templatePath);
+
+private :
+    /**
+     * @brief createParentDirectory Recursively tries to create the directory pointed to by path.
+     * @param path The directory path to be created.
+     * @return true on success, false on failure
```
Comment (breakreturn) [inline common/createdir.h]: Comment formatting is inconsistent with whitespace (compare this with the other comments).

File: common/createdir.h (hunk 26 above)
Comment (kursatkobya) [reply common/createdir.h]: The last six characters of template must be XXXXXX and these are replaced with a string that makes the directory name unique. This one is mentioned within comment in paramline area "@param templatePath a template Path used to create the path of the temporary directory,
     * including "XXXXXX" which will be replaced with a unique ID for the temporary directory"

File: common/unit-test/createdir_unittest.cpp [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+ */
+TEST_F(CreateDirTest, DISABLED_creationError)
+{
+    {
+        CreateDir tempCd;
+        ASSERT_TRUE(tempCd.createDirectory("/tmp/test/error/in/nested"));
+        chmod("/tmp/test/error/in/nested", 444);
+        ASSERT_FALSE(tempCd.createDirectory("/tmp/test/error/in/nested/directory/error"));
+    }
+    ASSERT_FALSE(isDirectory("/tmp/test/error/in/nested"));
```
Comment (tobsan) [inline common/unit-test/createdir_unittest.cpp]: Imho this type of check should be in all the tests to make sure that things are rolled back. Or it can be part of the fixture class.

**Review by tobsan — CHANGES_REQUESTED** [review]

**Review by tobsan — APPROVED** [review]
